Refactor EndpointInfo and ConnectionInfo in C++

bernardnormier opened the pull request:
This is the C++ equivalent of #3185 and #3187.

pepone commented on cpp/src/Ice/SSL/SecureTransportTransceiverI.cpp:
```diff
@@ -503,24 +503,22 @@ Ice::SSL::SecureTransport::TransceiverI::toDetailedString() const
     }
 
-    return info;
+    return make_shared<Ice::SSL::ConnectionInfo>(
```
A minor issue with this update, and other certificates. If the constructor throws, or `_delegate->getInfo()` throws. we leak the peerCertificate.

bernardnormier replied:
It's not clear to me why the constructor or _delegate->getInfo would throw.

pepone reviewed (approved):
Looks good, there was a failure with the MATLAB build.

bernardnormier commented:
> Looks good, there was a failure with the MATLAB build.

Small issue in the MATLAB code, now fixed. A more interesting issue was the failure on macos release, apparently due to weak vtables. Also fixed.
